modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -2911,6 +2912,27 @@ def setitem_builder(df, internal_indices=[]):  # pragma: no cover
             idx = self.get_axis(axis ^ 1).get_indexer_for([key])[0]
             return self.insert_item(axis ^ 1, idx, value, how, replace=True)
 
+        if axis == 0:
+            if hasattr(value, "dtype"):
+                value_dtype = value.dtype
+            elif is_scalar(value):
+                value_dtype = np.dtype(type(value))
+            else:
+                value_dtype = np.array(value).dtype
```

Comment:
It's probably better to move this into a separate function.

Reply:
will move for both `.insert()` and `__setitem()` after one of these PRs will be merged (#6757, #6758)

Reply:
#6757 is merged